lib/DependencyProvider.js

```diff
@@ -0,0 +1,367 @@
+// @flow
+
+const fs = require('fs');
+const os = require('os');
+const path = require('path');
+const wasm = require('elm-solve-deps-wasm');
+const SyncGet = require('./SyncGet.js');
+const collator = new Intl.Collator('en', { numeric: true }); // for sorting SemVer strings
+
+// Initialization work done only once.
+wasm.init();
+const syncGetWorker /*: {| get: (string) => string, shutDown: () => void |} */ =
+  SyncGet.startWorker();
+
+// Cache of existing versions according to the package website.
+let onlineVersionsCache /*: Map<string, Array<string>> */ = new Map();
+
+// Memoization cache to avoid doing the same work twice in listAvailableVersions.
+// This is to be cleared before each call to solve_deps().
+const listVersionsMemoCache /*: Map<string, Array<string>> */ = new Map();
+
+// Constructor of an offline solver.
+function newOffline() /*: {
```

Comment:
Why does this (and `newOnline`) return an object with the “real” function? Couldn’t we expose `solveOffline` which would be the internal `solve` function directly?

Reply:
You're totally right. I initially did this because I started with a class design but could not only expose what I wanted, all was public. So I replaced the class by a function returning an object. And then ended up removing all things I had put in the object except the actual useful function and kept the global variables of the module since it was more convenient.

In the end only the function stays so we can definitely peel that layer!